### PR TITLE
Next level config

### PR DESCRIPTION
**Pull request by benarmston**
Rework configuration

The new configuraiton aims to support simplified configuration with sensible convention / defaults.  In particular:

* `*.development{,.local}.yaml` config files which are only sourced when the app is being developed.

* Simplified support for paths being relative to `flight_ROOT` when installed with the openflight builder.  Including confguration files, paths to executables and paths to cron directories.

The mechanism used to achieve this is:

* Add a `Flight` module which has the methods `config`, `env`, `root` and `logger`.  This module will allow our common libraries, such as `flight_configuration` to access details about the app without requiring them to be injected.  This module is analogous to the `Rails` module.

* `Flight.env` defaults to `production` unless `FLIGHT_ENVIRONMENT` is set.  This ensures backwards compatibility at the expense of a little development inconvenience.

* The `root_path` set in the configuration is relative to `flight_ROOT` when `Flight.env` is production and `flight_ROOT` is set.  Otherwise it is relevant to the installation directory. This allows our configuration files and other paths to be relative to `/opt/flight` when deployed with an openflight builder, without having to maintain multiple configurations.

* A number of extensions to `FlightConfiguration::DSL` have been developed which need to be folded into `FlightConfiguration`.

### CODE REVIEW

**Review by WilliamMcCumstie — CHANGES_REQUESTED** [review]
This looks good for the most part, going to have a look at the build after this. However I have two major queries:

1. Assuming most of this will be extracted into a library, I dislike the dependency on `active_support`, and
2. I have a couple of reservations on how `flight_ENVIRONMENT` is set.

My various comments mostly concern `active_support`, so I won't rehash them any further.

---

Currently we default `flight_ENVIRONMENT` to `production`, which seems sensible at first but can lead to various issues.

Firstly, we don't want `development` environments to accidentally start using the production paths. This is particularly important for metadata files which will break the production version's validation when new keys are introduced.

Unfortunately, this can not be prevent by sticking `flight_ENVIRONMENT` in the `.bashrc` file. Doing so would be picked up by the production builds, and cause them to switch over to the development paths.

---

~The alternative is to make the `production` environment opt-in. This should alleviate the above issues.

All the production builds have `libexec/command` file which is ideal for setting the environment variable:~

```
export flight_ENVIRONMENT=production
export FLIGHT_PROGRAM_NAME='flight www'
export FLIGHT_CWD=$(pwd)
cd "${flight_ROOT}"/opt/www/cert
/opt/flight/bin/flexec bundle exec bin/cert "$@"
```

This means the underlying utility may do one of the following:
* Default to `development`, OR
* Raise an error saying it hasn't been set.

I am partial to raising an error as it means mistakes will be picked up in production.

File: Gemfile [hunk 1]
```diff
@@ -27,6 +27,7 @@
 #==============================================================================
 source 'https://rubygems.org'
 
+gem 'activesupport', require: 'active_support'
```
[the remark this answers is not in the document]
Comment (benarmston) [reply Gemfile:30]: I generally agree that libraries should be cautious about depending on `active_support`.  I don't intend to have `flight_configuration` depend on `active_support`.

Parts of this are intended to be extracted to a library and parts are to remain in the application.  All of the use of `active_support` is in the application side of things.

`flight_configuration` currently contains two modules.  `DSL` which provides mechanism and very little policy or convention.  `RackDSL` wraps `DSL` providing convention suitable for a rack application.  My intention is that `FcdslExtenstions` is to become a third module provided by `flight_configuration`.  That module wraps `DSL` and provides convention suitable for non-rack (cli only?) applications.

I would also imagine that `Configuration#from_config_file` and `Configuration#local_cofig_file` will be merged back to `flight_configuration`.

This new module will require a means to "hook" into the application.  That hook point is the new `Flight` module.  The `Flight` module is an application concern it contains methods to 1) get the configuration object; 2) get the environment; 3) get the root path; and 3) get a logger object.  If an application uses `active_support` in its `Flight` module, that's fine, it doesn't bleed into the library or into any other applications.

File: lib/flight_cert.rb [hunk 2]
```diff
@@ -91,6 +93,21 @@ def https_disabled?
...
+        ENV["flight_ENVIRONMENT"].presence || "production"
+      )
+    end
+
+    def root
+      @root ||=
+        if env.production? && ENV["flight_ROOT"].present?
+          File.expand_path(ENV["flight_ROOT"])
+        else
+          File.expand_path('..', __dir__)
```
[the remark this answers is not in the document]
Comment (benarmston) [reply lib/flight_cert.rb:107]: This method is not intended to be extracted into a library.  Instead it is used as a means for the library to query the application.  Our libraries will have a requirement that the application provides a `Flight` module and that the `Flight` module has a `root` method.

File: lib/flight_cert/configuration.rb [hunk 3]
```diff
@@ -32,6 +32,43 @@ module FlightCert
...
+      def application_name(name=nil)
+        @application_name ||= name
+        if @application_name.nil?
+          raise Error, 'The application_name has not been defined!'
+        end
+        @application_name
+      end
+
+      def root_path(*_)
+        super Flight.root
```
[the remark this answers is not in the document]
Comment (benarmston) [reply lib/flight_cert/configuration.rb:45]: `flight_configuration` does a similar thing in its `RackDSL` module [here](https://github.com/openflighthpc/flight_configuration/blob/master/lib/flight_configuration/rack_dsl.rb#L40) and [here](https://github.com/openflighthpc/flight_configuration/blob/master/lib/flight_configuration/rack_dsl.rb#L52).

Its fine when `RackDSL` does it as `RackDSL` is intended to provide convention over the mechanism provided by `DSL`.  This module will be extracted to another convention providing module.

File: lib/flight_cert/configuration.rb [hunk 4]
```diff
@@ -32,6 +32,43 @@ module FlightCert
...
+      end
+
+      def config_files(*_)
+        @config_files ||= [
+          root_path.join("etc/#{application_name}.yaml"),
+          root_path.join("etc/#{application_name}.#{Flight.env}.yaml"),
+          root_path.join("etc/#{application_name}.local.yaml"),
+          root_path.join("etc/#{application_name}.#{Flight.env}.local.yaml"),
```
[the remark this answers is not in the document]
Comment (benarmston) [reply lib/flight_cert/configuration.rb:53]: That very much depends on the app.  I'm not sure that we want that for admin tools.  Tools such as `flight-job` we very much do.  `flight-job`'s configuration will have to be written something like:

```ruby
class Configuration
  extend FlightConfiguration::CliDSL

  config_files "~/.config/flight/<application_name>.yaml"
end
```

I'm not entirely sold on that.  Perhaps `extra_config_files "~/...."` would read more clearly.  I think that decision can wait until `flight-job` is updated.

**Comment by benarmston** [comment]
@WilliamMcCumstie I've added responses to some of your inline comments.

Regarding the usage of `active_support`, I don't intend for it to end up in `flight_configuration`.  I can't see that having an application depend on `active_support` is a problem.  If the install size ever becomes an issue, it should be simple to change the application code to remove `active_support`.  Requiring only specific parts of `active_support` (instead of the entire thing), should not noticeably affect boot times.

**Review by WilliamMcCumstie — APPROVED** [review]

**Comment by WilliamMcCumstie** [comment]
Yep the application level dependency on `active_support` is fine, just wanted to make sure it wasn't going to be ported.

File: Gemfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply Gemfile:30]: :+1:

File: lib/flight_cert/configuration.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply lib/flight_cert/configuration.rb:53]: :+1:

File: lib/flight_cert/configuration.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply lib/flight_cert/configuration.rb:45]: Ah far point

File: lib/flight_cert.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply lib/flight_cert.rb:107]: :+1:

**Pull request merged**